propertyestimator/properties/enthalpy.py

```diff
@@ -2,154 +2,290 @@
 A collection of enthalpy physical property definitions.
 """
 
-from propertyestimator.properties.plugins import register_estimable_property
+from collections import namedtuple
+
+from simtk import unit
+
 from propertyestimator.datasets.plugins import register_thermoml_property
+from propertyestimator.properties.plugins import register_estimable_property
 from propertyestimator.properties.properties import PhysicalProperty
+from propertyestimator.substances import Mixture
 from propertyestimator.thermodynamics import Ensemble
+from propertyestimator.utils.serialization import PolymorphicDataType
 from propertyestimator.utils.statistics import AvailableQuantities
-from propertyestimator.workflow import WorkflowSchema
-from propertyestimator.workflow import protocols, groups
-from propertyestimator.workflow.utils import ProtocolPath
+from propertyestimator.workflow import WorkflowSchema, plugins, groups
+from propertyestimator.workflow import protocols
+from propertyestimator.workflow.decorators import protocol_input, protocol_output
+from propertyestimator.workflow.schemas import ProtocolReplicator
+from propertyestimator.workflow.utils import ProtocolPath, PlaceholderInput
+
+
+@plugins.register_calculation_protocol()
+class WeightValueByMoleFraction(protocols.BaseProtocol):
+    """Multiplies a value by the mole fraction of a component
+    in a mixture substance.
+    """
+    @protocol_input(unit.Quantity)
+    def value(self, value):
```

Comment:
This is my first time digging into this convention. It's a little weird to me that we have decorators+functions being defined for what a (badly-abused) dict could do... I have to look especially closely at this in particular because it's is the interface that future developers will use when Simon transitions to a more science-based role. Maybe @andrrizzi could also give his thoughts? I won't make this a blocking review, but let's chat about this on our call Monday.

Reply:
100% agree with the phrase badly-abused dict. A dict in this case would absolutely and unequivocally be a bad design choice here. These inputs and outputs are very much attributes of the class they exist in, and should be defined as such. Aside from this, putting these in a dict would be a use-ability and maintenance nightmare - off the top of my head quickly:

- for one there would be no IDE integration. When people are using these classes the IDE would not be able to suggest which inputs and outputs are available to get / set, making them a pain to use without having to dig through the docstring.
- you can't define doc strings for things in a dict. You would have to put them either in the __init__ docstring (bad choice) or in the class docstring where they might be hard to locate.
- all attributes would be have to be accessed by string literals both internally in the class and externally by the user. Not only does this result in really ugly looking code, it is just asking for typos leading to access bugs especially given that, as far as I'm aware, most static linters will not check if a key is present in a dict.
- It becomes awkward to tag attributes with meta, such as their anticipated types or their merge behavior, without either creating new dicts to store this in, or having the value dicts store class which have the value and meta data inside, which again will most likely lead to ugly code.

While the current descriptor design pattern may look a bit odd (there may be another way to clean this up), it avoids the above issues and gives a significant amount of control over how these attributes are `get` and `set` (such as validation type checking), allows easy metadata attachment, and in future will allow us to do some interesting, automatic stuff with docstrings.

Reply:
I'm a big fan of descriptors, and I never thought of using decorators to add the docstring!

I'd design this a little backwards though, to make this route of adding the descriptor with `__get__` and `__set__` possible
```python
class MyClass:
    value = ProtocolInput(unit.Quantity)
    value.__doc__ = "blabla"

    weighted_value = ProtocolOutput(unit.Quantity)
    weighted_value.__doc__ = "The value weighted by the `component`s mole fraction as determined from the `full_substance`."
```
which I think it's a slightly little more transparent. You can still have a decorator like
```python
def protocol_input(value_type, merge_behavior=MergeBehaviour.ExactlyEqual):
    def _protocol_input(cls_attribute):
        desc = ProtocolInput(value_tupe, merge_behavior)
        descr.__doc__ = cls_attribute.__doc__
        return descr
    return _protocol_input
```
Also, not sure if you're aware of the special function `__set_name__()`, which is pretty handy to determine the private name of the instance attribute in descriptors: https://stackoverflow.com/questions/42023852/how-can-i-get-the-attribute-name-when-working-with-descriptor-protocol-in-python